File: src/hunt_optimizer/persistence/HuntMethodPersister.ts
import { Persister } from "../../core/persistence";
import { Server } from "../../core/model";
import { HuntMethodModel } from "../model/HuntMethodModel";
import { Duration } from "luxon";

const METHOD_USER_TIMES_KEY = "HuntMethodStore.methodUserTimes";

class HuntMethodPersister extends Persister {
    persist_method_user_times(hunt_methods: HuntMethodModel[], server: Server): void {
        const user_times: PersistedUserTimes = {};

        for (const method of hunt_methods) {
            if (method.user_time.val != undefined) {
                user_times[method.id] = method.user_time.val.as("hours");
            }
        }

        this.persist_for_server(server, METHOD_USER_TIMES_KEY, user_times);
    }

    async load_method_user_times(hunt_methods: HuntMethodModel[], server: Server): Promise<void> {
        const user_times = await this.load_for_server<PersistedUserTimes>(
            server,
            METHOD_USER_TIMES_KEY,
        );

        if (user_times) {
            for (const method of hunt_methods) {
                const hours = user_times[method.id];
                method.set_user_time(
                    hours == undefined ? undefined : Duration.fromObject({ hours }),
                );
            }
        }
    }
}

type PersistedUserTimes = { [method_id: string]: number };

export const hunt_method_persister = new HuntMethodPersister();

File: web/src/main/kotlin/world/phantasmal/web/huntOptimizer/persistence/HuntMethodPersister.kt
package world.phantasmal.web.huntOptimizer.persistence

import world.phantasmal.web.core.models.Server
import world.phantasmal.web.core.persistence.KeyValueStore
import world.phantasmal.web.core.persistence.Persister
import world.phantasmal.web.huntOptimizer.models.HuntMethodModel
import kotlin.time.Duration
import kotlin.time.DurationUnit.HOURS

class HuntMethodPersister(keyValueStore: KeyValueStore) : Persister(keyValueStore) {
    suspend fun persistMethodUserTimes(huntMethods: List<HuntMethodModel>, server: Server) {
        val userTimes = mutableMapOf<String, Double>()

        for (method in huntMethods) {
            method.userTime.value?.let { userTime ->
                userTimes[method.id] = userTime.toDouble(HOURS)
            }
        }

        persistForServer(server, METHOD_USER_TIMES_KEY, userTimes)
    }

    suspend fun loadMethodUserTimes(huntMethods: List<HuntMethodModel>, server: Server) {
        loadForServer<Map<String, Double>>(server, METHOD_USER_TIMES_KEY)?.let { userTimes ->
            for (method in huntMethods) {
                userTimes[method.id]?.let { userTime ->
                    method.setUserTime(Duration.hours(userTime))
                }
            }
        }
    }

    companion object {
        private const val METHOD_USER_TIMES_KEY = "HuntMethodStore.methodUserTimes"
    }
}
